### PR TITLE
Move beefy-merkle-tree to utils/binary-merkle-tree and make it generic

### PR DESCRIPTION
**Pull request by DaviRain-Su**
close: #13075 

We made our own `beefy light client` (https://github.com/octopus-network/beefy-light-client) for the library that we also want to use in the near contract, so we want to keep `beefy-merkle-tree` from introducing any substrate crate here. near contract has not been compiled.

polkadot companion: https://github.com/paritytech/polkadot/pull/6528

### CODE REVIEW

**Review by acatangiu — DISMISSED** [review]
I don't understand the purpose of this PR. Can you please add more details to it and/or it's associated issue?
Please describe the problem and the proposed solution.

**Comment by DaviRain-Su** [comment]
> I don't understand the purpose of this PR. Can you please add more details to it and/or it's associated issue? Please describe the problem and the proposed solution.

Added some background description information

**Review by serban300 — COMMENTED** [review]
I think the general idea makes sense, but it would be hard to move everything apart from the algorithmic part outside of the crate. For example the `sp_api::decl_runtime_apis! {` part is usually kept in the primitives (not sure why). Anyway, I can think of 2 options:

1. Moving just the algorithmic part in a different crate that wouldn't have many dependencies. But seems a bit too much.
2. Conditional compilation. Defining a target that would compile only the algorithmic logic and would bring with it only the required dependencies.

Another thing related to this. This algorithmic logic seems to be more related to mmr than to beefy. So maybe it would be worth moving it in `primitives/merkle-mountain-range` as well. I'm not sure. Anyway, not a requirement. If it makes sense, we can do it later as part of a different PR.

**Comment by acatangiu** [comment]
>     1. Moving just the algorithmic part in a different crate that wouldn't have many dependencies. But seems a bit too much.
> 
>     2. Conditional compilation. Defining a target that would compile only the algorithmic logic and would bring with it only the required dependencies.

Both options 1 & 2 look good to me, somewhat leaning toward option 1.
It seems that the only "non-algorithmic logic" is the `decl_runtime_apis!` which I added here at some point, but in retrospect it's not really related with the rest of the crate.
So `decl_runtime_apis` can move to the pallet (like this PR does) or if that doesn't work, to its own separate `primitives/beefy-mmr` crate.

"The algorithmic logic" is used by the `beefy-mmr` pallet for:
1. building the MMR leaves that are later used by BEEFY in consensus (all BEEFY basically does is run consensus on the MMR root - thus the leaves' contents),
2. building a BEEFY `ValidatorSet` as a very compact `BeefyAuthoritySet`: it contains just the merkle root of the tree built from validator keys as leaves.

So it is related to both BEEFY and MMR. But that's just how _we_ use it. It can be used in many ways, so I agree it could go in its own crate independent of `pallet-beefy`, `pallet-mmr` or `pallet-beefy-mmr` (it will just be used by one/more of them).

On that note, I would also change its name to something without `beefy`, and change its location to `utils/` or `primitives/`.

**Comment by DaviRain-Su** [comment]
Because this `ValidationContext` trait is pub, it seems that `MsgEnvelope` must also be pub.
```rust
  pub trait ValidationContext {
        /// Validation entrypoint.
        fn validate(&self, message: MsgEnvelope) -> Result<(), RouterError>
        where
            Self: Sized,
        {
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2238385

**Review by acatangiu — COMMENTED** [review]
Overall seems ok to me. Please also make the changes required for tests as well and I will review it all then.

**Comment by acatangiu** [comment]
> On that note, I would also change its name to something without beefy, and change its location to `utils/` or `primitives/`.

@paritytech/sdk-node wdyt? any other suggestion?

File: utils/merkle-tree/src/lib.rs [hunk 1]
```diff
@@ -30,40 +30,73 @@
...
+extern crate alloc;
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+/// Supported hashing output size.
+///
+/// The size is restricted to 32 bytes to allow for a more optimised implementation.
+pub type Hash = [u8; 32];
```
Comment (bkchr) [inline utils/merkle-tree/src/lib.rs:43]: This doesn't make any sense? What does "more optimized implementation" mean? You only use this for the `Keccak256` hasher below.

File: utils/merkle-tree/src/lib.rs [hunk 2]
```diff
@@ -30,40 +30,73 @@
...
+mod keccak256 {
+	use tiny_keccak::{Hasher as _, Keccak};
+	/// Keccak256 hasher implementation.
+	pub struct Keccak256;
+	impl Keccak256 {
+		/// Hash given data.
+		pub fn hash(data: &[u8]) -> super::Hash {
+			<Keccak256 as super::Hasher>::hash(data)
+		}
+	}
```
Comment (bkchr) [inline utils/merkle-tree/src/lib.rs:55]: In general you seem to try to make this as generic as possible, but then you add this hasher here? I would remove this. This isn't that much code, that you can just reimplement where you need it. For beefy we can put it into the beefy primitive crate.

File: utils/merkle-tree/Cargo.toml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "beefy-merkle-tree"
+name = "merkle-tree"
```
Comment (bkchr) [inline utils/merkle-tree/Cargo.toml:2]:
```suggestion
name = "binary-merkle-tree"
```

Should be more expressive. First time I was thinking on what the difference to our trie is until I read the description.

**Comment by bkchr** [comment]
> > On that note, I would also change its name to something without beefy, and change its location to `utils/` or `primitives/`.
> 
> @paritytech/sdk-node wdyt? any other suggestion?

Sounds okay to me.

File: utils/merkle-tree/src/lib.rs (hunk 2 above)
Comment (DaviRain-Su) [reply utils/merkle-tree/src/lib.rs]: have remove `Keccak256` module.

File: utils/merkle-tree/src/lib.rs (hunk 1 above)
Comment (DaviRain-Su) [reply utils/merkle-tree/src/lib.rs]: Removed

File: utils/merkle-tree/Cargo.toml (hunk 3 above)
Comment (DaviRain-Su) [reply utils/merkle-tree/Cargo.toml]: updated

**Comment by acatangiu** [comment]
@DaviRain-Su please pull master and rebase this PR

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2257371

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2257372

File: Cargo.toml [hunk 4]
```diff
@@ -246,6 +245,7 @@ members = [
 	"utils/frame/rpc/client",
 	"utils/prometheus",
 	"utils/wasm-builder",
+	"utils/merkle-tree",
```
Comment (acatangiu) [inline Cargo.toml:248]: for clarity the path could be named same as crate:
```suggestion
	"utils/binary-merkle-tree",
```

**Review by bkchr — APPROVED** [review]

File: utils/binary-merkle-tree/src/lib.rs [hunk 5]
```diff
@@ -287,20 +289,20 @@ where
...
 	visitor: &mut V,
-) -> Result<H::Output, Vec<H::Output>>
+) -> Result<H::Out, Vec<H::Out>>
 where
-	H: HashT,
-	H::Output: AsRef<[u8]> + PartialOrd,
-	V: Visitor<H::Output>,
-	I: Iterator<Item = H::Output>,
+	H: Hasher,
+	H::Out: AsRef<[u8]> + PartialOrd + PartialOrd,
```
Comment (serban300) [inline utils/binary-merkle-tree/src/lib.rs:297]: Nit: `+ PartialOrd` is added twice here.

File: utils/binary-merkle-tree/Cargo.toml [hunk 6]
```diff
@@ -11,20 +11,15 @@ homepage = "https://substrate.io"
...
 
 [features]
 debug = ["array-bytes", "log"]
 default = ["debug", "std"]
-std = [
-	"beefy-primitives/std",
-	"sp-api/std",
-	"sp-runtime/std"
-]
+std = []
```
Comment (serban300) [inline utils/binary-merkle-tree/Cargo.toml:25]: Should we add
```
    "log/std",
    "hash-db/std"
```

here ?

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot rebase

**Comment by bkchr** [comment]
@DaviRain-Su please merge master.

**Comment by bkchr** [comment]
bot merge

**Comment by DaviRain-Su** [comment]
I don't know how to fix.

**Review by serban300 — APPROVED** [review]

**Comment by serban300** [comment]
Doesn't seem to be a problem with the PR. It's probably the CI. Not sure how to fix it myself, but I expect it to be fixed soon.

**Comment by acatangiu** [comment]
bot merge

**Pull request merged**